File: CousinsInABinaryTree.java
// Time Complexity :O(n)
// Space Complexity : O(h) worstcase - O(n)
// Did this code successfully run on Leetcode : Yes https://leetcode.com/problems/cousins-in-binary-tree/submissions/1062708286/
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach in three sentences only
/*
 * Maintain 2 variables -- x_found, y_found
 * Do a bfs traversal, check the root's left and right values matches with x and y or viceversa accordinly. 
 * if yes return false -- which concludes both are not from different parents
 * Check if current root value is same as x or y, if yes update the  x_found, y_found values, if any of the values is false return false
 */
import java.util.*;
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        boolean x_found = false;
        boolean y_found = false;
        if(root == null)
            return false;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int size = q.size();
            for(int i=0;i<size;i++){
                TreeNode curr = q.poll();
                if(curr.left!=null && curr.right!=null){
                    if(curr.left.val == x && curr.right.val == y)
                        return false;
                    if(curr.left.val == y && curr.right.val == x)
                        return false;
                }
                if(curr.val == x)
                    x_found = true;
                if(curr.val ==y)
                    y_found = true;
                if(curr.left!=null)
                    q.add(curr.left);
                if(curr.right!=null)
                    q.add(curr.right);
           
            }
            if(x_found == true && y_found == true)
                return true;
            if(x_found == true || y_found == true)
                return false;
        }
        
        return true;

    }
}

File: cousinsInABinaryTree.py
# // Time Complexity :O(N)
# // Space Complexity : O(D) --> leaf nodes for BFS, O(H)--> height for DFS
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach

##we do a normal level order traversal and if come across a node whose vsalue matches with x or y with update the found and parent
## in the end of level we check if xfound and yfound and then compare their parents
class solution:
    def cousinsBFS(self, root, x,y):

        queue = collections.deque()

        queue.append(root, None)

        while queue:
            qsize = len(queue)
            x_found = False
            y_found = False
            x_parent = None
            y_parent = None
            for i in range(qsize):
                curr_node, parent = queue.popleft()

                if curr_node.val == x:
                    x_found = True
                    x_parent = parent
                if curr_node.val == y:
                    y_found = True
                    y_parent = parent
                
                if curr_node.left:
                    queue.append([curr_node.left, curr_node])
                    queue.append([curr_node.right, curr_node])
                
            if x_found and y_found:
                return x_parent != y_parent

            if x_found or y_found:
                return False

        return False    
## here we do the bfs without considering the parents. so instead of calculating the validity and parents after putting them in the queue
## we will check if they are not cousins and return 
    def cousinsBFS_without_parent(self, root, x, y):

        queue = collections.deque()
        queue.append(root)

        while queue:
            qsize = len(queue)
            x_found = False
            y_found = False

            for i in range(qsize):
                curr = queue.popleft()

                if curr.val == x:
                    xfound = True
                
                if curr.val ==y:
                    yfound = True
                
                if curr.left and curr.right:
                    if curr.left.val == x and curr.right.val == y:
                        return False
                    if curr.left.val == y and curr.right.val == x:
                        return False
                
                if curr.left:
                    queue.append(curr.left)
                if curr.right:
                    queue.append(curr.right)
            if x_found and y_found:
                return True
            if x_found or y_found: return False
        
        return False
## In dfs we maintain a the depth along with parents. at the end of the dfs we see if depth is same and parents are different.
    def cousinsDFS(self, root, x, y):
        self.x_parent = None
        self.y_parent = None
        self.x_depth = -1
        self.y_depth =-2
        self.helper(root,x,y,0,None)
        return self.x_depth == self.y_depth and self.y_parent == self.x_parent
    def helper(self, root, x,y, depth, parent):
        ## base
        if root ==None: return
        ##logic

        if root.val == x:
            self.x_depth = depth
            self.x_parent = parent
        if root.val == y:
            self.y_depth=depth
            self.y_parent = parent
        
        if self.x_parent == None or self.y_parent == None:
            self.helper(root.left,x,y,depth+1, root)
        if self.x_parent == None or self.y_parent == None:
            self.helper(root.right, x, y, depth+1, root)
        



